src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm

```diff
@@ -499,79 +525,95 @@ - (void)_pruneEmptyStoredAttributesBranches
+        NSArray<NSNumber *> * endpointIndex = [self _fetchEndpointIndexForNodeID:nodeID];
+        NSMutableArray<NSNumber *> * endpointIndexCopy = endpointIndex.mutableCopy;
 
         for (NSNumber * endpointID in endpointIndex) {
             // Fetch endpoint index
-            NSMutableArray<NSNumber *> * clusterIndex = [self _fetchClusterIndexForNodeID:nodeID endpointID:endpointID].mutableCopy;
-            NSUInteger clusterIndexCount = clusterIndex.count;
+            NSArray<NSNumber *> * clusterIndex = [self _fetchClusterIndexForNodeID:nodeID endpointID:endpointID].mutableCopy;
+            NSMutableArray<NSNumber *> * clusterIndexCopy = clusterIndex.mutableCopy;
 
             for (NSNumber * clusterID in clusterIndex) {
                 // Fetch endpoint index
-                NSMutableArray<NSNumber *> * attributeIndex = [self _fetchAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID].mutableCopy;
-                NSUInteger attributeIndexCount = attributeIndex.count;
+                NSArray<NSNumber *> * attributeIndex = [self _fetchAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID].mutableCopy;
+                NSMutableArray<NSNumber *> * attributeIndexCopy = attributeIndex.mutableCopy;
 
                 for (NSNumber * attributeID in attributeIndex) {
                     NSDictionary * value = [self _fetchAttributeValueForNodeID:nodeID endpointID:endpointID clusterID:clusterID attributeID:attributeID];
 
                     if (!value) {
-                        [attributeIndex removeObject:attributeID];
+                        [attributeIndexCopy removeObject:attributeID];
                     }
                 }
 
-                if (!attributeIndex.count) {
-                    [clusterIndex removeObject:clusterID];
-                } else if (attributeIndex.count != attributeIndexCount) {
-                    BOOL success = [self _storeAttributeIndex:attributeIndex forNodeID:nodeID endpointID:endpointID clusterID:clusterID];
+                if (attributeIndex.count != attributeIndexCopy.count) {
+                    BOOL success;
                     if (!success) {
+                        if (attributeIndexCopy.count) {
+                            success = [self _storeAttributeIndex:attributeIndexCopy forNodeID:nodeID endpointID:endpointID clusterID:clusterID];
+                        } else {
+                            [clusterIndexCopy removeObject:clusterID];
+                            success = [self _deleteAttributeIndexForNodeID:nodeID endpointID:endpointID clusterID:clusterID];
+                        }
                         storeFailures++;
-                        MTR_LOG_INFO("Store failed for attributeIndex");
+                        MTR_LOG_INFO("Store failed in _pruneEmptyStoredAttributesBranches for attributeIndex @ 0x%016llX:0x%04X:0x%08lX", nodeID.unsignedLongLongValue, endpointID.unsignedShortValue, clusterID.unsignedLongValue);
```

Comment:
Again, decimal endpoints.  Will stop commenting on that; just fix all of them, please.